src/v/cluster/archival_metadata_stm.h

```diff
@@ -20,31 +21,50 @@
 
 #include <seastar/util/log.hh>
 
+#include <system_error>
+
 namespace cluster {
 
+namespace details {
+/// This class is supposed to be implemented in unit tests.
+class archival_metadata_stm_accessor;
+} // namespace details
+
 /// This replicated state machine allows storing archival manifest (a set of
 /// segments archived to cloud storage) in the archived partition log itself.
 /// This is needed to 1) avoid querying cloud storage on partition startup and
 /// 2) to replicate metadata to raft followers so that they can decide which
 /// segments can be safely evicted.
 class archival_metadata_stm final : public persisted_stm {
+    friend class details::archival_metadata_stm_accessor;
+
 public:
     explicit archival_metadata_stm(
       raft::consensus*, cloud_storage::remote& remote, ss::logger& logger);
 
-    /// Add the difference between manifests to the raft log, replicate it and
+    /// Add segments to the raft log, replicate them and
     /// wait until it is applied to the STM.
     ss::future<std::error_code> add_segments(
       const cloud_storage::partition_manifest&,
```

Comment:
> The add_segments method now still accepts the partition_manifest but it's used as a collection of segments. This is done to avoid including cloud_storage/partition_manifest.h in the cluster/archival_metadata_stm.h directly.

hmm, so c++ doesn't allow using `const std::vector<segment_meta>&` with only a fwd declaration of `segment_meta`?

Reply:
If I'll change parameter from partition_manifest (which has forward declaration in `cloud_storage/fwd.h`) to `const std::vector<cloud_storage::segment_meta>&` I'll have to include `cloud_storage/types.h`. I'm trying to avoid this. The `partition_manifest` is used as a collection of segments here.

Reply:
got it 👍